File: src/epay_tool/ai/provider.py
"""Provider-agnostic AI interface.

The app never talks to a specific model directly - it talks to this
interface. Swapping Gemini for a local Ollama model (or any future
provider) is a config change, not a refactor.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from pathlib import Path

from epay_tool.ai.schemas import ListingCopy, ListingReview, ProductIdentification
from epay_tool.config import Settings


class AIProvider(ABC):
    @abstractmethod
    def identify_product(
        self,
        photos: list[Path],
        known: dict[str, str] | None = None,
        notes: str | None = None,
    ) -> ProductIdentification:
        """Identify a product from photos; ask questions where uncertain."""

    @abstractmethod
    def write_copy(self, facts: dict) -> ListingCopy:
        """Generate title (<=80 chars) and HTML description from known facts."""

    @abstractmethod
    def review_listing(self, draft: dict) -> ListingReview:
        """Soft quality review of a complete draft before posting."""


class NullProvider(AIProvider):
    """Used when AI is disabled - the app still works for barcode matches."""

    def identify_product(self, photos, known=None, notes=None) -> ProductIdentification:
        return ProductIdentification(
            questions=["KI ist deaktiviert - bitte Produktdaten manuell eintragen."]
        )

    def write_copy(self, facts) -> ListingCopy:
        title = str(facts.get("title") or facts.get("product_name") or "")[:80]
        return ListingCopy(title=title, description_html=f"<p>{title}</p>")

    def review_listing(self, draft) -> ListingReview:
        return ListingReview()


def get_provider(settings: Settings) -> AIProvider:
    name = settings.ai_provider.lower()
    if name == "gemini":
        from epay_tool.ai.gemini import GeminiProvider

        return GeminiProvider(settings)
    if name == "ollama":
        from epay_tool.ai.ollama import OllamaProvider

        return OllamaProvider(settings)
    return NullProvider()

File: src/main/ai/provider.ts
/**
 * Provider-agnostic AI interface.
 *
 * The app never talks to a specific model directly - it talks to this
 * interface. Swapping Gemini for a local Ollama model (or any future
 * provider) is a config change, not a refactor. Nothing outside this
 * directory imports a vendor SDK.
 */

import type { AppConfig } from '@main/config'

import type { ListingCopy, ListingReview, ProductIdentification } from './schemas'

export interface IdentifyInput {
  /** Absolute paths of local photos. */
  photoPaths: string[]
  /** Anything already known, e.g. the barcode. */
  known?: Record<string, string>
  /** Free text the seller added. */
  notes?: string | null
}

export interface AIProvider {
  /** Identify a product from photos; ask questions where uncertain. */
  identifyProduct(input: IdentifyInput): Promise<ProductIdentification>
  /** Generate title (<=80 chars) and HTML description from known facts. */
  writeCopy(facts: Record<string, unknown>): Promise<ListingCopy>
  /** Soft quality review of a complete draft before posting. */
  reviewListing(draft: Record<string, unknown>): Promise<ListingReview>
}

/** Used when AI is disabled - the app still works for barcode matches. */
export class NullProvider implements AIProvider {
  async identifyProduct(): Promise<ProductIdentification> {
    return {
      aspects: {},
      questions: ['KI ist deaktiviert – bitte Produktdaten manuell eintragen.'],
      photoSuggestions: [],
    }
  }

  async writeCopy(facts: Record<string, unknown>): Promise<ListingCopy> {
    const title = String(facts.title ?? facts.productName ?? '').slice(0, 80)
    return { title, descriptionHtml: `<p>${title}</p>` }
  }

  async reviewListing(): Promise<ListingReview> {
    return { issues: [], suggestions: [] }
  }
}

export interface ProviderResult {
  provider: AIProvider
  /** Why the configured provider is unavailable, for the UI to surface. */
  error: string | null
}

/**
 * Never throws. A missing API key must not stop the app from starting:
 * barcode lookups go through eBay's catalog and work without any AI at
 * all, so the app degrades to that instead of refusing to launch.
 */
export async function createProvider(config: AppConfig): Promise<ProviderResult> {
  try {
    switch (config.aiProvider) {
      case 'gemini': {
        const { GeminiProvider } = await import('./gemini')
        return { provider: new GeminiProvider(config), error: null }
      }
      case 'ollama': {
        const { OllamaProvider } = await import('./ollama')
        return { provider: new OllamaProvider(config), error: null }
      }
      default:
        return { provider: new NullProvider(), error: null }
    }
  } catch (error) {
    return {
      provider: new NullProvider(),
      error: error instanceof Error ? error.message : String(error),
    }
  }
}
